clang/lib/Driver/Driver.cpp

```diff
@@ -4786,11 +4786,20 @@ class OffloadingActionBuilder final {
           return ABRT_Success;
 
         Action *DeviceCompilerInput = nullptr;
-        for (Action *&A : SYCLDeviceActions) {
+        for (auto TargetActionInfo :
+             llvm::zip(SYCLDeviceActions, SYCLTargetInfoList)) {
+          Action *&A = std::get<0>(TargetActionInfo);
+          auto &TargetInfo = std::get<1>(TargetActionInfo);
           types::ID OutputType = types::TY_LLVM_BC;
           if ((SYCLDeviceOnly || Args.hasArg(options::OPT_emit_llvm)) &&
               Args.hasArg(options::OPT_S))
             OutputType = types::TY_LLVM_IR;
+          if (SYCLDeviceOnly &&
+              TargetInfo.TC->getTriple().getEnvironment() ==
+                  llvm::Triple::SYCLMLIR &&
+              Args.hasArg(options::OPT_emit_mlir)) {
+            OutputType = types::TY_MLIR_IR;
+          }
```

Comment:
If a source file is compiled using options `-fsycl-device-only -emit-mlir` we should emit MLIR code. Why do we also need to check that the triple is  'SYCLMLIR' ?

Reply:
Fixed